### PR TITLE
Move inlay hints implementation in core

### PR DESCRIPTION
**Pull request by yyoncho**

### CODE REVIEW

File: lsp-mode.el [hunk 1]
```diff
@@ -1965,7 +1966,7 @@ regex in IGNORED-FILES."
     lsp-actionscript lsp-ada lsp-angular lsp-astro lsp-bash lsp-beancount lsp-clangd
     lsp-clojure lsp-cmake lsp-crystal lsp-csharp lsp-css lsp-d lsp-dhall
     lsp-dockerfile lsp-elixir lsp-elm lsp-erlang lsp-eslint lsp-fortran lsp-fsharp lsp-gdscript
-    lsp-go lsp-gleam lsp-graphql lsp-groovy lsp-hack lsp-haxe lsp-html lsp-idris lsp-javascript lsp-json lsp-kotlin lsp-lua
+    lsp-go lsp-gleam lsp-graphql lsp-groovy lsp-hack lsp-haxe lsp-html lsp-idris lsp-mode lsp-json lsp-kotlin lsp-lua
```
Comment (jcs090218) [inline lsp-mode.el:1969]: This shouldn't be renamed to `lsp-mode`.

Comment (yyoncho) [reply lsp-mode.el:1969]: Ah, thank you. I was wondering why lsp-javascript is not loaded. :)

File: lsp-mode.el [hunk 2]
```diff
@@ -9457,6 +9463,110 @@ defaults to `progress-bar."
...
+  "Mode for displaying inlay hints."
+  :lighter nil
+  (cond
+   (lsp-inlay-hints-mode
+    (add-hook 'lsp-on-idle-hook #'lsp--update-inlay-hints nil t)
+    (when lsp-update-inlay-hints-on-scroll
+      (add-to-list (make-local-variable 'window-scroll-functions)
+                   #'lsp--update-inlay-hints-scroll-function)))
+   (t
+    (lsp--remove-overlays 'lsp-javascript-inlay-hint)
```
Comment (geza-herman) [inline lsp-mode.el:9565]: This should be `'lsp-inlay-hint`

File: lsp-protocol.el [hunk 3]
```diff
@@ -791,7 +785,15 @@ See `-let' for a description of the destructuring mechanism."
  (UnregistrationParams (:unregisterations) nil)
  (WatchKind nil nil)
  (WillSaveTextDocumentParams (:reason :textDocument) nil)
- (WorkspaceSymbolParams (:query) nil))
+ (WorkspaceSymbolParams (:query) nil)
+ ;; 3.17
+ (InlayHint (:label :position :kind) (:paddingLeft :paddingRight?))
```
Comment (geza-herman) [inline lsp-protocol.el:790]: I had to rename this to `:paddingRight` (without the `?`), otherwise right padding didn't work

**Review by geza-herman — COMMENTED** [review]
I hope you don't mind, I found two little possible bugs

File: lsp-protocol.el (hunk 3 above)
Comment (yyoncho) [reply lsp-protocol.el:790]: Yeah, it was present in the previous implementation.

**Review by ericdallo — COMMENTED** [review]
awesome!

File: lsp-mode.el [hunk 4]
```diff
@@ -7700,7 +7706,7 @@ read-only location, you can instruct lsp-mode to use them via
 
  (eval-after-load `lsp-mode
    `(progn
-      (require lsp-javascript)
+      (require lsp-mode)
```
Comment (geza-herman) [inline lsp-mode.el:7709]: I think this is also an accidental rename

**Comment by kiennq** [comment]
@yyoncho Can we go with this PR? I saw all the test passed already.

**Review by jcs090218 — APPROVED** [review]
LGTM too! So +1. :D

**Comment by ylc** [comment]
I am using this with clangd. Works great !!! Thank you very much.

**Comment by yyoncho** [comment]
> Is there anything special I might need to do to get inlay hints working with `go-mode` and `gopls`?
> 
> I've I've had no luck after the following:
> 
> ```emacs-lisp
> (setq lsp-inlay-hint-enable t)
> 
> (lsp-register-custom-settings
>  '(("gopls.hints" ((assignVariableTypes . t)
>                    (compositeLiteralFields . t)
>                    (compositeLiteralTypes . t)
>                    (constantValues . t)
>                    (functionTypeParameters . t)
>                    (parameterNames . t)
>                    (rangeVariableTypes . t)))))
> ```
> 
> I've verified that `lsp-inlay-hints-mode` is enabled in Go buffers, and with the `-rpc.trace` flag passed to `gopls`, I can see `textDocument/inlayHint` requests and responses going through.
> 
> But no hints show in the any Go buffers. While hints do now work Rust buffers.

@jimeh my guess is that it is missing a flag.

**Comment by yyoncho** [comment]
> I am using this with clangd. Works great !!! Thank you very much.

@jimeh most likely fixed after https://github.com/emacs-lsp/lsp-mode/pull/4044

**Pull request merged**